### ISSUE TITLE
udev rules syntax error

### ISSUE THREAD

**Issue by martinpitt** [systemd/systemd#4728]
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> master (not in 232)

Seen while experimenting with master builds:
```
Setting up udev (232-135-gfd79a95-0) ...
Error, MATCH expected but got ' evdev:atkbd:dmi:bvn*:bvr*:bd*:svnHewlett-Packard*:pn*[pP][aA][vV][iI][lL][iI][oO][nN]*13*x360*:pvr*' in '/lib/udev/hwdb.d/60-keyboard.hwdb':
Error, MATCH expected but got ' KEYBOARD_KEY_d7=unknown' in '/lib/udev/hwdb.d/60-keyboard.hwdb':
```

Will fix later, just want this as a reminder as I'm working on something else  ATM.

**Comment by martinpitt** [comment]
This is a bit embarassing:
```
SKIP: hwdb/parse_hwdb.py
```
in both semaphore and autopkgtest. Both are missing the `pyparsing` module, which is easy enough to add, but then it fails bitterly:
```
FAIL: hwdb/parse_hwdb
=====================

WARNING: evdev is not available
Cannot parse ./hwdb/60-evdev.hwdb: Expected {{{{{{Combine:({{{"mouse:" {"usb" ^ "bluetooth" ^ "ps2" ^ "*"} ":"} ^ {"keyboard:" {"name"} ":"} ^ {"touchpad:" {"i8042" ^ "rmi" ^ "bluetooth" ^ "usb"} ":"} ^ {"evdev:" {"name" ^ "atkbd" ^ "input"} ":"}} W:(0123...)}) Suppress:(LineEnd)} ^ Suppress:({Python style comment Suppress:(LineEnd)})}}... - {{{Suppress:(<SPC>) Combine:({{{W:(ABCD...,ABCD...) - "="} - W:(ABCD...)} - [Python style comment]}) Suppress:(LineEnd)} ^ Suppress:({<SPC> Python style comment Suppress:(LineEnd)})}}...} - Suppress:({{LineStart LineEnd} ^ stringEnd})} ^ {Suppress:({{Python style comment Suppress:(LineEnd)}}...) - Suppress:({LineStart LineEnd})}} (at char 0), (line:1, col:1)
./hwdb/60-evdev.hwdb: 0 match groups, 0 matches, 0 properties
Cannot parse ./hwdb/60-keyboard.hwdb: Expected {{{{{{Combine:({{{"mouse:" {"usb" ^ "bluetooth" ^ "ps2" ^ "*"} ":"} ^ {"keyboard:" {"name"} ":"} ^ {"touchpad:" {"i8042" ^ "rmi" ^ "bluetooth" ^ "usb"} ":"} ^ {"evdev:" {"name" ^ "atkbd" ^ "input"} ":"}} W:(0123...)}) Suppress:(LineEnd)} ^ Suppress:({Python style comment Suppress:(LineEnd)})}}... - {{{Suppress:(<SPC>) Combine:({{{W:(ABCD...,ABCD...) - "="} - W:(ABCD...)} - [Python style comment]}) Suppress:(LineEnd)} ^ Suppress:({<SPC> Python style comment Suppress:(LineEnd)})}}...} - Suppress:({{LineStart LineEnd} ^ stringEnd})} ^ {Suppress:({{Python style comment Suppress:(LineEnd)}}...) - Suppress:({LineStart LineEnd})}} (at char 0), (line:1, col:1)
./hwdb/60-keyboard.hwdb: 0 match groups, 0 matches, 0 properties
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pyparsing.py", line 2139, in parseFile
    file_contents = file_or_filename.read()
AttributeError: 'str' object has no attribute 'read'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./hwdb/parse_hwdb.py", line 195, in <module>
    groups = parse(fname)
  File "./hwdb/parse_hwdb.py", line 136, in parse
    parsed = grammar.parseFile(fname)
  File "/usr/lib/python3/dist-packages/pyparsing.py", line 2142, in parseFile
    file_contents = f.read()
  File "/usr/lib/python3.5/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 14592: ordinal not in range(128)
FAIL hwdb/parse_hwdb.py (exit status: 1)
```

I'll have a look at that, as this is really the kind of thing we should be able to detect automatically in PRs.

**Comment by martinpitt** [comment]
I fixed the UnicodeDecodeError and the syntax error in `hwdb/60-keyboard.hwdb`, but I still find the other error messages if `parse_hwdb` completely impossible to interpret -- there are no line numbers or line contents, and `systemd-hwdb update` is perfectly able to read those.

**Comment by martinpitt** [comment]
I gave up on `parse_hwdb.py` for now. It isn't able to read most of our hwdb files (we only call it on 60-* and 70-*), the error messages are unintelligible, and it complains about valid files. I'll file a PR which adds a new simple test that just calls `systemd-hwdb update` in the built tree in a temporary directory.

@keszybz, do you remember if `parse_hwdb.py` ever actually worked? We should either fix it to work on the current hwdb, or drop it entirely, as having a broken version in the tree that never gets run in CI doesn't help anyone.

**Comment by keszybz** [comment]
It seems to work well enough here:
```
$ hwdb/parse_hwdb.py
hwdb/70-pointingstick.hwdb: 6 match groups, 16 matches, 8 properties
hwdb/70-mouse.hwdb: 63 match groups, 92 matches, 72 properties
hwdb/60-evdev.hwdb: 30 match groups, 43 matches, 110 properties
hwdb/70-touchpad.hwdb: 3 match groups, 5 matches, 3 properties
Cannot parse hwdb/60-keyboard.hwdb: Expected stringEnd (at char 18062), (line:418, col:1)
hwdb/60-keyboard.hwdb: 0 match groups, 0 matches, 0 properties
1
```
That was before 54abf90a0b, which fixes line 419. The error reporting is far from perfect, I know, but fixing it would require much more coding. Current stuff in the hwdb passes the test, so the idea is that if you introduce any breaking change, and hwdb_test.py starts failing, you just have to look at the new lines.

**Comment by martinpitt** [comment]
@keszybz : Interesting; I tried it again with pyparsing 2.0.3 (in Ubuntu 16.04) and that does work. But it's completely broken with 2.1.8 (in Ubuntu 16.10 and Debian sid). So we actually could run this in CI as that runs on 16.04 usually, but I can't add it as a build dependency as then we could not build it in Debian/Ubuntu any more. But good to know that this actually works, thanks!

**Comment by keszybz** [comment]
On Sat, Nov 26, 2016 at 09:56:32AM -0800, Martin Pitt wrote:
> @keszybz : Interesting; I tried it again with pyparsing 2.0.3 (in Ubuntu 16.04) and that does work. But it's completely broken with 2.1.8 (in Ubuntu 16.10 and Debian sid). So we actually could run this in CI as that runs on 16.04 usually, but I can't add it as a build dependency as then we could not build it in Debian/Ubuntu any more. But good to know that this actually works, thanks!

If it doesn't work with some version of pyparsing, please file a bug,
I'll try to fix it.

Here I have pyparsing 2.1.5, and that seems to work fine, but it's
of course possible that something changed between 2.1.5 and 2.1.8.

**Comment by whot** [comment]
can confirm this works here with 2.1.5 as well, in both py2 and py3 (I get the "Warnign: evdev is not available" but meh). Also, for the archives: this was reverted in commit 2f08ccc again because the encoding breaks on python2

**Comment by martinpitt** [comment]
I tested this again: pyparsing 2.1.8 is still fine, 2.1.10 is broken. https://bugs.debian.org/844368 reports another regression with 2.1.9/2.1.10, so I think `parse_hwdb.py` is fine after all, and it's a pyparsing problem. I'll follow it there for the time being.

Sorry for the confusion!

**Comment by martinpitt** [comment]
Mystery solved, see PR #4772
